File: Stack/20.valid-parentheses.go
package leetcode

// @leet start
func isValid(s string) bool {
	var (
		stack []rune
		dict  = map[rune]rune{
			')': '(',
			'}': '{',
			']': '[',
		}
	)
	for _, r := range s {
		if res, ok := dict[r]; !ok {
			// find left pair enqueue.
			stack = append(stack, r)
		} else { // try cancel pair
			n := len(stack)
			if n == 0 { // no item
				return false
			}
			if res != stack[n-1] {
				return false
			}
			stack = stack[:n-1] // pop
		}
	}
	return len(stack) == 0
}

// @leet end

File: TwoPointer/20.valid-parentheses.py
# @leet start
class Solution:
    def isValid(self, s: str) -> bool:
        map = {
            "(": ")",
            "{": "}",
            "[": "]",
        }
        st = []
        for c in s:
            if c in map:
                st.append(map[c])
            else:
                if len(st) == 0 or st.pop() != c:
                    return False
        return len(st) == 0


# @leet end

